Add protocol handlers to _WKApplicationManifest

szewai opened the pull request:
#### 6a56145baf5a3ef3752d3766353cd8638ab93c07
<pre>
Add protocol handlers to _WKApplicationManifest
<a href="https://bugs.webkit.org/show_bug.cgi?id=265486">https://bugs.webkit.org/show_bug.cgi?id=265486</a>
<a href="https://rdar.apple.com/106139265">rdar://106139265</a>

Reviewed by NOBODY (OOPS!).

Parse protocol handlers from Web Application Manifest and pass it to WebKit clients.
Spec: <a href="https://wicg.github.io/manifest-incubations/#protocol_handlers-member">https://wicg.github.io/manifest-incubations/#protocol_handlers-member</a>
Test: ApplicationManifestParserTest.ProtocolHandlers
      ApplicationManifest.Coding

* Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::supportedSchemes): We currently only support &quot;mailto&quot; as we have concern on other schemes, see discussion in
radar.
(WebCore::normalizedParametersAndCreateProtocolHandler):
(WebCore::isInScope):
(WebCore::ApplicationManifestParser::parseManifest):
(WebCore::ApplicationManifestParser::parseProtocolHandlers):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h: We use NSString for url because NSURL encodes characters,
(&quot;%s&quot; token will become &quot;%25s&quot;) which might be confusing for clients when constructing URL for request (by replacing
&quot;%s&quot; token with actual identifier).
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(makeVectorElement):
(+[_WKApplicationManifestProtocolHandler supportsSecureCoding]):
(-[_WKApplicationManifestProtocolHandler initWithCoder:]):
(-[_WKApplicationManifestProtocolHandler protocol]):
(-[_WKApplicationManifestProtocolHandler url]):
(-[_WKApplicationManifestProtocolHandler initWithCoreProtocolHandler:]):
(-[_WKApplicationManifestProtocolHandler encodeWithCoder:]):
(-[_WKApplicationManifestProtocolHandler dealloc]):
(-[_WKApplicationManifest initWithCoder:]):
(-[_WKApplicationManifest encodeWithCoder:]):
(-[_WKApplicationManifest protocolHandlers]):
* Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
(ApplicationManifestParserTest::testProtocolHandlers):
(TEST_F):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ApplicationManifest.mm:
(TestWebKitAPI::TEST):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a56145baf5a3ef3752d3766353cd8638ab93c07

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28264 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6907 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29622 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30794 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25755 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28761 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8949 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4287 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26048 "Build was cancelled. Recent messages:Failed to print configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; layout-tests (exception)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28532 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5683 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24329 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4945 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5076 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25327 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31483 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25886 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25761 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31376 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5046 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3227 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29130 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6627 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5483 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5564 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
```diff
@@ -109,6 +121,7 @@ - (instancetype)initWithCoder:(NSCoder *)coder
     _sizes = adoptNS([[coder decodeObjectOfClasses:[NSSet setWithArray:@[[NSArray class], [NSString class]]] forKey:@"sizes"] copy]);
     _type = adoptNS([[coder decodeObjectOfClass:[NSString class] forKey:@"type"] copy]);
     _purposes = adoptNS([[coder decodeObjectOfClasses:[NSSet setWithArray:@[[NSArray class], [NSNumber class]]] forKey:@"purposes"] copy]);
+    _purposes = adoptNS([[coder decodeObjectOfClasses:[NSSet setWithArray:@[[NSArray class], [NSNumber class]]] forKey:@"purposes"] copy]);
```
Nit - this looks redundant.

szewai replied:
ah nice catch! I removed it and added encoding test case for for protocol_handlers.

szewai committed:
Add protocol handlers to _WKApplicationManifest
https://bugs.webkit.org/show_bug.cgi?id=265486
rdar://106139265

Reviewed by NOBODY (OOPS!).

Parse protocol handlers from Web Application Manifest and pass it to WebKit clients.
Spec: https://wicg.github.io/manifest-incubations/#protocol_handlers-member
Test: ApplicationManifestParserTest.ProtocolHandlers
      ApplicationManifest.Coding

* Source/WebCore/Modules/applicationmanifest/ApplicationManifest.h:
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::supportedSchemes): We currently only support "mailto" as we have concern on other schemes, see discussion in
radar.
(WebCore::normalizedParametersAndCreateProtocolHandler):
(WebCore::isInScope):
(WebCore::ApplicationManifestParser::parseManifest):
(WebCore::ApplicationManifestParser::parseProtocolHandlers):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.h: We use NSString for url because NSURL encodes characters,
("%s" token will become "%25s") which might be confusing for clients when constructing URL for request (by replacing
"%s" token with actual identifier).
* Source/WebKit/UIProcess/API/Cocoa/_WKApplicationManifest.mm:
(makeVectorElement):
(+[_WKApplicationManifestProtocolHandler supportsSecureCoding]):
(-[_WKApplicationManifestProtocolHandler initWithCoder:]):
(-[_WKApplicationManifestProtocolHandler protocol]):
(-[_WKApplicationManifestProtocolHandler url]):
(-[_WKApplicationManifestProtocolHandler initWithCoreProtocolHandler:]):
(-[_WKApplicationManifestProtocolHandler encodeWithCoder:]):
(-[_WKApplicationManifestProtocolHandler dealloc]):
(-[_WKApplicationManifest initWithCoder:]):
(-[_WKApplicationManifest encodeWithCoder:]):
(-[_WKApplicationManifest protocolHandlers]):
* Tools/TestWebKitAPI/Tests/WebCore/ApplicationManifestParser.cpp:
(ApplicationManifestParserTest::testProtocolHandlers):
(TEST_F):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ApplicationManifest.mm:
(TestWebKitAPI::TEST):
